enderio-machines/src/main/java/com/enderio/machines/common/blockentity/AlloySmelterBlockEntity.java

```diff
@@ -1,260 +1,333 @@
+            .outputSlot()
+            .build();
+    }
+
+    private boolean acceptSlotInput(int slot, ItemStack stack) {
+        // Ensure we don't break automation by inserting items that'll break the current recipe.
+        PoweredCraftingTask<IAlloySmeltingRecipe, IAlloySmeltingRecipe.Container> currentTask = getCurrentTask();
+        if (currentTask != null) {
+            MachineInventory inventory = getInventory();
+            ItemStack currentContents = inventory.getStackInSlot(slot);
+            inventory.setStackInSlot(slot, stack);
+
+            boolean accept = currentTask.getRecipe().matches(container, level);
+
+            inventory.setStackInSlot(slot, currentContents);
+            return accept;
+        }
+        return true;
+    }
+
+    @Override
+    protected void onInventoryContentsChanged(int slot) {
+        inventoryChanged = true; // TODO: 28/05/2022 This kind of thing might be a good idea for a base crafter class?
+        super.onInventoryContentsChanged(slot);
+    }
+
+    @Override
+    protected boolean hasNextTask() {
+        return inventoryChanged;
+    }
+
+    @Override
+    protected @Nullable PoweredCraftingTask<IAlloySmeltingRecipe, IAlloySmeltingRecipe.Container> getNextTask() {
+        if (level == null)
+            return null;
+
+        // Mark inventory changed false again
+        inventoryChanged = false;
+
+        // Search for an alloy recipe.
+        if (getMode().canAlloy()) {
+            var task = level
+                .getRecipeManager()
+                .getRecipeFor(MachineRecipes.Types.ALLOY_SMELTING, container, level)
+                .map(this::createTask)
+                .orElse(null);
+
+            if (task != null)
+                return task;
+        }
+
+        // Search for a smelting recipe.
+        if (getMode().canSmelt()) {
+            var task = level
+                .getRecipeManager()
+                .getRecipeFor(RecipeType.SMELTING, getRecipeWrapper(), level)
+                .map(recipe -> createTask(new VanillaAlloySmeltingRecipe(recipe)))
+                .orElse(null);
+
+            if (task != null)
+                return task;
+        }
+
+        return null;
+    }
+
+    @Override
+    protected @Nullable PoweredCraftingTask<IAlloySmeltingRecipe, IAlloySmeltingRecipe.Container> loadTask(CompoundTag nbt) {
+        PoweredCraftingTask<IAlloySmeltingRecipe, IAlloySmeltingRecipe.Container> task = createTask(null);
+        task.deserializeNBT(nbt);
+        return task;
+    }
+
+    private PoweredCraftingTask<IAlloySmeltingRecipe, IAlloySmeltingRecipe.Container> createTask(@Nullable IAlloySmeltingRecipe recipe) {
+        return new PoweredCraftingTask<>(energyStorage, recipe, level, container) {
+
+            @Override
+            protected void takeInputs(IAlloySmeltingRecipe recipe) {
+                // TODO: Maybe delegate this to the recipe by passing in the MachineRecipe itself? Or just the container?
```

Comment:
I'd love some feedback on this in particular. Unsure quite how to handle machine inventory handler.